scripts/train_lm.py

```diff
@@ -149,65 +149,101 @@ def FormatMetaparameters(metaparameters, num_train_sets):
-        # Note: the first #num_train_sets weights are not needed to be checked (they can be the same)
-        if i < num_train_sets:
-            out_param.append(format(out[i]))
-        else:
-            x = out[i]
-            for j in range(i + 1, size):
-                y = out[j]
-                # we will round to the next digit when we detect repeated values
-                round_digit = round_index + 1
-                round_max_digit = max(len(str(metaparameters[i])) - 2, len(str(metaparameters[j])) -2)
-                while x == y and round_digit <= round_max_digit:
-                    x = round(metaparameters[i], round_digit)
-                    y = round(metaparameters[j], round_digit)
-                    round_digit += 1
-                # actually validate_metaparameters.py will validate those parameters later. 
-                # except for satisfying condition that those parameters should be in 
-                # range (0, 1), parameters D1, D2, D3, D4 for a certain xgram model 
-                # should satisfy D1 > D2 > D3 > D4. 
-                # here we only checked the repeating values but not the order relationship
-                if x == y:
-                    sys.exit("train_lm.py: {0} and {1} are the same. metaparameters can not be exactly the same.".format(x,y))
-                out[j] = y
-        
-            # out[i] now is valided and can be appended to the output set out_param 
-            out[i] = x
-            out_param.append(format(out[i]))
+                sys.exit("train_lm.py: {0}th parameter is {1}. metaparameters should be in range (0, 1)".format(i + 1, out[i]))
     
+    # append the first #num_train_sets parameters to the output set 
+    # note: the first #num_train_sets parameters (not data-type weights) do not need to be checked
+    assert num_train_sets >= 1 and num_train_sets < len(metaparameters)
+    for param in out[:num_train_sets]:
+        out_param.append(format(param))
+   
+    # check repeating value in the rest parameters. 
+    # after this while loop, there should be no repeating values in set out
+    (index1, index2) = FindRepeatingValue(out[num_train_sets:])
+    while (index1 != -1):
+        i = index1 + num_train_sets
+        j = index2 + num_train_sets
+        (updated_param1, updated_param2) = AdjustValues(out[i], out[j], i, j, metaparameters[i], metaparameters[j])
+        out[i] = updated_param1
+        out[j] = updated_param2
+        (index1, index2) = FindRepeatingValue(out[num_train_sets:])
+
+    # now parameters are validated and can be appended to the output set
+    for param in out[num_train_sets:]: 
+        out_param.append(format(param))
+
     return ','.join(out_param)
 
+# this function checks whether there are repeating values in parameters(excluding 
+# the first #number numbers, which are not data-type weights) and if found, return
+# the index of the first pair repeating values 
+def FindRepeatingValue(parameters):
```

Comment:
I think the work for excluding the first numbers is not done inside the function. You may need to move that line of comment to the caller.